pkg/storegateway/indexheader/reader_pool.go

```diff
@@ -87,34 +90,59 @@ func (l lazyLoadedHeadersSnapshot) persist(persistDir string) error {
-func NewReaderPool(logger log.Logger, lazyReaderEnabled bool, lazyReaderIdleTimeout time.Duration, metrics *ReaderPoolMetrics, lazyLoadedSnapshotConfig LazyLoadedHeadersSnapshotConfig) *ReaderPool {
-	p := newReaderPool(logger, lazyReaderEnabled, lazyReaderIdleTimeout, metrics)
+func NewReaderPool(logger log.Logger, lazyReaderEnabled bool, lazyReaderIdleTimeout time.Duration, eagerLoadIndexReaderEnabled bool, metrics *ReaderPoolMetrics, lazyLoadedSnapshotConfig LazyLoadedHeadersSnapshotConfig) *ReaderPool {
+	var snapshot *lazyLoadedHeadersSnapshot
+	if lazyReaderEnabled && eagerLoadIndexReaderEnabled {
+		lazyLoadedSnapshotFileName := filepath.Join(lazyLoadedSnapshotConfig.Path, lazyLoadedHeadersListFile)
+		var err error
+		snapshot, err = loadLazyLoadedHeadersSnapshot(lazyLoadedSnapshotFileName)
+		if err != nil {
+			level.Warn(logger).Log("msg", "loading lazy-loaded index header failed; removing the file", "file", lazyLoadedSnapshotFileName, "err", err)
+			err := os.Remove(lazyLoadedSnapshotFileName)
+			if err != nil {
+				level.Warn(logger).Log("msg", "removing lazy-loaded index header failed", "file", lazyLoadedSnapshotFileName, "err", err)
+			}
+		}
+	}
+
+	p := newReaderPool(logger, lazyReaderEnabled, lazyReaderIdleTimeout, eagerLoadIndexReaderEnabled, metrics, snapshot)
 
 	// Start a goroutine to close idle readers (only if required).
 	if p.lazyReaderEnabled && p.lazyReaderIdleTimeout > 0 {
 		checkFreq := p.lazyReaderIdleTimeout / 10
 
 		go func() {
-			tickerLazyLoadPersist := time.NewTicker(time.Minute)
-			defer tickerLazyLoadPersist.Stop()
-
 			tickerIdleReader := time.NewTicker(checkFreq)
 			defer tickerIdleReader.Stop()
 
-			for {
-				select {
-				case <-p.close:
-					return
-				case <-tickerIdleReader.C:
-					p.closeIdleReaders()
-				case <-tickerLazyLoadPersist.C:
-					snapshot := lazyLoadedHeadersSnapshot{
-						IndexHeaderLastUsedTime: p.LoadedBlocks(),
-						UserID:                  lazyLoadedSnapshotConfig.UserID,
+			if p.eagerLoadReaderEnabled {
+				tickerLazyLoadPersist := time.NewTicker(time.Minute)
+				defer tickerLazyLoadPersist.Stop()
+
+				for {
+					select {
+					case <-p.close:
+						return
+					case <-tickerIdleReader.C:
+						p.closeIdleReaders()
+					case <-tickerLazyLoadPersist.C:
+						snapshot := lazyLoadedHeadersSnapshot{
+							IndexHeaderLastUsedTime: p.LoadedBlocks(),
+							UserID:                  lazyLoadedSnapshotConfig.UserID,
+						}
+
+						if err := snapshot.persist(lazyLoadedSnapshotConfig.Path); err != nil {
+							level.Warn(p.logger).Log("msg", "failed to persist list of lazy-loaded index headers", "err", err)
+						}
 					}
-
-					if err := snapshot.persist(lazyLoadedSnapshotConfig.Path); err != nil {
-						level.Warn(p.logger).Log("msg", "failed to persist list of lazy-loaded index headers", "err", err)
+				}
+			} else {
+				for {
+					select {
+					case <-p.close:
+						return
+					case <-tickerIdleReader.C:
+						p.closeIdleReaders()
```

Comment:
To reduce the duplication here, what if we had one `select` statement with all three channels being read, but only initialise `tickerLazyLoadPersist` if `p.eagerLoadReaderEnabled` is `true`? (`select`-ing a `nil` channel is safe and the channel will be ignored.)

Reply:
`tickerLazyLoadPersist` has to be always initialized, otherwise it will be nil and `tickerLazyLoadPersist.C` will fail with nil pointer. That's why I duplicated most of the select blocks previously.

What I try to do instead is top stop ticker immediately if we know we don't need it. I feel it strange and might not be working. But let me know what do you think.

```
			tickerLazyLoadPersist := time.NewTicker(time.Minute)
			defer tickerLazyLoadPersist.Stop()
			if !p.eagerLoadReaderEnabled {
				// if eager load of lazy load index headers is not enabled, just stop the ticker now
				tickerLazyLoadPersist.Stop()
			}
```

Reply:
Couldn't we do something like this?

```go
var lazyLoadC <-chan time.Time

if p.eagerLoadReaderEnabled {
	tickerLazyLoadPersist := time.NewTicker(time.Minute)
	defer tickerLazyLoadPersist.Stop()

	lazyLoadC = tickerLazyLoadPersist.C
}
```

And then `select` on `lazyLoadC`

Reply:
I think I have been thinking too hard and didn't realise that easy way. Thank you. Updated as suggested.